src/merkle.rs

```diff
@@ -49,25 +49,45 @@ pub const BUILD_DATA_BLOCK_SIZE: usize = 64 * BUILD_CHUNK_NODES;
 /// will be nil.
 ///
 /// TODO: Ord
+///
+/// This tree structure can consist of at most 3 layers of trees (of
+/// arity U, N and R, from bottom to top).
+///
+/// U is the arity of the sub-trees.
+/// N is the number of sub-trees that are present (arity of second layer trees).
+/// R is the arity of the top level tree.
+///
+/// With N and R defaulting to 1, the tree should perform as a legacy
+/// merkle tree without layers.
+#[allow(clippy::type_complexity)]
 #[derive(Debug, Clone, Eq, PartialEq, Default)]
-pub struct MerkleTree<T, A, K, U = U2>
+pub struct MerkleTree<T, A, K, U = U2, N = U0, R = U0>
```

Comment:
Soooo, I know we have been using single letters for these generics, but I think we should start giving them more descriptive names, given the amount we have now. for example

```rust
pub struct MerkleTree<Element, Algorithm, Store, SubTreeArity = U2, NumSubTrees = U0, TopTreeArity = U0> {
}
```

Reply:
I like this :+1:

Reply:
Although I think I'd use `BaseTreeArity`, `SubTreeArity`, and `TopTreeArity`.

Reply:
yeah feel free to choose better names, used these just to demonstrate what I meant

Reply:
Took a stab at this, but Element/Algorithm/Store clash with the trait names, so I renamed those to E, A and S for now.